Feature/tandem favorites TM-1739 TM-1815

scott062 opened the pull request:
To user test this ->
DASHBOARD
- Favorites Dashboard should show correct tabs and count for User_1 AP/PV tabs and Tandem_2 AP/PV tabs
- Dashboard tandem results cards should be "limited" for tandem tabs (No Submit Bid Option)
- Dashboard tandem results cards should be styled for tandem 2 (No artifact tandem_1 styled favorites)
- Default error message for each tab if there are no favorites (tab specific now)

SEARCH
- Tandem search results page is going to compare all `tandem_nbr: 1` positions with normal favorites, but all `tandem_nbr: 2` results will be compared against the tandem favorites. Adding a tandem_nbr 1 position to favorites will only update that card and not the corresponding tandem 2 duplicate. You will see that recently added `tandem_nbr_: 1` position in NORMAL favorites and all `tandem_nbr: 2` favorites in TANDEM favorites

scott062 commented on src/Components/Favorite/Favorite.jsx:
```diff
@@ -62,10 +62,12 @@ class Favorite extends Component {
+    const { compareArray, refKey, isTandem, isTandemTwo } = nextProps;
     const oldState = this.getSavedState();
-    const newState = existsInArray(refKey, compareArray);
+    const newState = isTandem ?
+      existsInArray(refKey, compareArray, isTandem, isTandemTwo)
+      :
+      existsInArray(refKey, compareArray);
```
This might not be necessary since existsInArray sets defaults for isTandem and isTandem two. This is just an extra useless check

scott062 commented on src/Components/Favorite/Favorite.jsx:
```diff
@@ -76,8 +78,11 @@ class Favorite extends Component {
     // Is the refKey in the array? If so, return true
-    const { compareArray, refKey } = this.props;
-    return existsInArray(refKey, compareArray);
+    const { compareArray, refKey, isTandem, isTandemTwo } = this.props;
+    return isTandem ?
+      existsInArray(refKey, compareArray, isTandem, isTandemTwo)
+      :
+      existsInArray(refKey, compareArray);
```
Same as above comment

scott062 commented on src/Components/ResultsCard/ResultsCard.jsx:
```diff
@@ -160,19 +163,31 @@ class ResultsCard extends Component {
       refKey: result.id,
     };
 
+    if (isProjectedVacancy && isTandem) {
+      options.favorite.compareArray = favoritesPVTandem;
+    } else if (isProjectedVacancy) {
+      options.favorite.compareArray = favoritesPV;
+    } else if (isTandem) {
+      options.favorite.compareArray = favoritesTandem;
+    } else {
+      options.favorite.compareArray = favorites;
+    }
+
```
Is there a cleaner syntax for this?

elizabeth-jimenez replied:
maybe:
`if (isTandem) {
    options.favorite.compareArray = isProjectedVacancy ? favoritesPVTandem : favoritesTandem;
} else {
    options.favorite.compareArray = isProjectedVacancy ? favoritesPV : favorites;
}`

featuring the ternary operator we both love ❤️

scott062 commented on src/utilities.js:
```diff
@@ -239,10 +239,20 @@ export const shortenString = (string, shortenTo = 250, suffix = '...') => {
 // for checking if a favorite_position exists in the user's profile
-export const existsInArray = (ref, array) => {
+export const existsInArray = (ref, array, isTandem = false, isTandemTwo = false) => {
   let found = false;
   array.forEach((i) => {
-    if (get(i, 'id') && ref && `${i.id}` === `${ref}`) {
+    if (isTandem) {
+      // Two possible types of arrays for tandem check
+      // Search Results compares an array of objects {id: #, tandem: t/f}
+      // Favorite List View compares an array of pos objects { id: #,...,tandem_nbr: 1/2}
+      // Last condition in following 'if' statment is to account for the two comparison arrays
+      const nbr = isTandemTwo ? 2 : 1;
+      if (get(i, 'id') && ref && `${i.id}` === `${ref}` &&
+      (get(i, 'tandem') === isTandemTwo || get(i, 'tandem_nbr') === nbr)) {
+        found = true;
+      }
+    } else if (get(i, 'id') && ref && `${i.id}` === `${ref}`) {
       found = true;
     }
   });
```
🙈

mjoyce91 commented:
@scott062 is this still WIP?

scott062 commented:
> @scott062 is this still WIP?

@mjoyce91 Should be good

mjoyce91 commented:
@scott062 having two issues
- Available tandem positions get shown as added and the button changes state on the results page, but don't appear on the Favorites page
- Projected vacancy positions show up on the Favorites page, but the button state doesn't change once they're added from the results page

scott062 commented:
> @scott062 having two issues
> 
> * Available tandem positions get shown as added and the button changes state on the results page, but don't appear on the Favorites page
> * Projected vacancy positions show up on the Favorites page, but the button state doesn't change once they're added from the results page

@mike should be good now

mjoyce91 commented:
@scott062 - take a look at conflicts

mjoyce91 commented:
![Screen Shot 2020-07-21 at 9 30 54 AM](https://user-images.githubusercontent.com/11576347/88060965-e5a8bc00-cb34-11ea-8f53-a25e9bfcccb0.png)

Clicking one of these should append the "tandem=true" param so that the bid list button isn't displayed on the details page

mjoyce91 commented:
When you only have tandem Favorites, they get included in the count but don't appear on the dashboard

![Screen Shot 2020-07-21 at 10 03 05 AM](https://user-images.githubusercontent.com/11576347/88064550-6ff31f00-cb39-11ea-997b-0c84da9f870c.png)

mjoyce91 commented:
Above comment could also apply to the Homepage Favorites

scott062 commented:
> ![Screen Shot 2020-07-21 at 9 30 54 AM](https://user-images.githubusercontent.com/11576347/88060965-e5a8bc00-cb34-11ea-8f53-a25e9bfcccb0.png)
> 
> Clicking one of these should append the "tandem=true" param so that the bid list button isn't displayed on the details page

Fixed this and fixed the add/remove button with correct comparison array on position details page

mjoyce91 commented:
Pagination does not display for Tandem PV favorites

![Screen Shot 2020-07-21 at 10 21 12 AM](https://user-images.githubusercontent.com/11576347/88066608-f27cde00-cb3b-11ea-80ee-c114e398ceba.png)

elizabeth-jimenez commented:
Cases: 

- [ ] No favorites -> adding one ap-favorite non-tandem-search:  good
- [ ] Adding only more than one ap-favorite non-tandem-search:  good
- [ ] No favorites -> adding one pv-favorite non-tandem-search: good
- [ ] Adding only more than one pv-favorite non-tandem-search: good

- [ ] No favorites -> adding one ap-favorite to tandem 1 on tandem-search: good
- [ ] Adding only more than one ap-favorite to tandem 1 on tandem-search: good
- [ ] No favorites -> adding one pv-favorite to tandem 1 on tandem-search: good
- [ ] Adding only more than one pv-favorite to tandem 1 on tandem-search: good

- [ ] No favorites -> adding one ap-favorite to tandem 2 on tandem-search: need to make sure ‘Add To Favorites’ is ‘Remove from Favorites’ and ‘Add to Bid List’ is hidden: good (fixed in update)
- [ ] Adding only more than one ap-favorite to tandem 2 on tandem-search: good
- [ ] No favorites -> adding one pv-favorite to tandem 2 on tandem-search: Button not updating to ‘Remove from Favorites’ after adding favorite (both on card in Search Results and in Position Details)
- [ ] Adding only more than one pv-favorite to tandem 2 on tandem-search: Button not updating to ‘Remove from Favorites’ after adding favorite (both on card in Search Results and in Position Details)

- [ ] Same ap Position added to favorites for TandemUser1 and TandemUser2:  same as above, make sure that once it is removed from tandem1 user favorites, when the positions details is accessed via tandem2 user favorite, the button says ‘Remove from Favorites’). Also, when going to details page from tandemUser2 and clicking ‘Remove from Favorites’, it is removed from TandemUser1: good (fixed in update)
- [ ] Same pv Position added to favorites for TandemUser1 and TandemUser2: Button not updating to ‘Remove from Favorites’ after adding favorite (both on card in Search Results and in Position Details

elizabeth-jimenez commented:
![image](https://user-images.githubusercontent.com/23065001/88075211-79cf4f00-cb46-11ea-81e9-60cb3060befd.png)
Maybe update the wording to be tandem favorites specific?

mjoyce91 commented:
@scott062 I'm still having the same issue adding PV Tandem favorites on the results page

scott062 commented:
> @scott062 I'm still having the same issue adding PV Tandem favorites on the results page

@mjoyce91 Just found the bug - should work now

mjoyce91 commented:
Looks good on my end. Failing tests should just be snapshots.

elizabeth-jimenez commented:
@scott062 just one more thing! - [ ] Undo hit in Toast Notification sends the Favorite to the right place (Tandem1 AP, Tandem1 PV, Tandem2 AP, Tandem2 PV)

scott062 commented:
> @scott062 just one more thing!
> 
> * [ ]  Undo hit in Toast Notification sends the Favorite back to the right place (Tandem1 AP, Tandem1 PV, Tandem2 AP, Tandem2 PV)

@elizabeth-jimenez Fixed
